docs/source/user_guide.md

```diff
@@ -153,8 +165,190 @@ name: new-project
+
+```
+> $env:BAR = 'set-in-shell'
+> conda project activate
+```
+
+## Defining and running commands
+
+Conda Project supports running commands as if the desired environment were activated, similar
+to [conda run](https://docs.conda.io/projects/conda/en/latest/commands/run.html).
+
+In Conda Project the `run` command can be used to execute ad-hoc commands or those defined
+in the `conda-project.yml` file. Note that `conda project run` should not be utilized after
+`conda project activate`. The `run` command will not work from within an activated environment.
+
+Let's start with a simple python script called `vars.py`, that prints environment variables
+and optionally takes an argument `--version` to also print the Python version.
+
+```python
+import os
+import sys
+
+if len(sys.argv) > 1 and sys.argv[1] == '--version':
+    print("0.0.1")
+    sys.exit(1)
+else:
+    print(f"The value of FOO is {os.environ.get('FOO')}")
+    print(f"The value of BAR is {os.environ.get('BAR')}")
+    print(f"The value of BAZ is {os.environ.get('BAZ')}")
 ```
 
+Note that when the `--version` flag is provided the return code for this script is 1, meaning
+a failure.
+
+We can execute this script with the `default` environment as an ad-hoc command using the Python
+interpreter provided by the default environment. Here a `.env` file is utilized to provide the
+value of the BAR variable.
+
+```
+> BAR=set-on-cli conda project run python vars.py
+The value of FOO is has-default-value
+The value of BAR is set-on-cli
+The value of BAZ is None
+```
+
+On Linux, Mac, and Windows the return code of the `conda project run` is set to the
+return code of the command it is executing. For example the above invocation returns
+0, known as a successful execution. The `--version` flag, however exits with code 1, typically
+meaning a failed execution.
+
+```
+> BAR=set-on-cli conda project run python vars.py --version
+0.0.1
+> echo $?
+1
+```
+
+On Windows `cmd.exe` we can echo the `%ERRORLEVEL% variable.
+
+```
+> set BAR=set-on-cli
+> conda project run python vars.py --version
+> echo %ERRORLEVEL%
+1
+```
+
+And in Powershell we print the `$LASTEXITCODE` variable
+
+```
+> $env:BAR = 'set-on-cli'
+> conda project run python vars.py --version
+0.0.1
+> $LASTEXITCODE
+1
+```
+
+Defined commands in the conda-project.yml are placed under the `commands:` key. Commands
+written in one line are set to execute over the default (first) environment.
+
+```yaml
+commands:
+  print-vars: python vars.py
+```
+
+Now you can use `conda project run` without arguments and it will execute the first
+named command in the `conda-project.yml` file.
+
+```
+> BAR=set-on-cli conda project run
+The value of FOO is has-default-value
+The value of BAR is set-on-cli
+The value of BAZ is None
+```
+
+Defined commands also support extra arguments, but the name of the command must be supplied.
+
+```
+> BAR=set-on-cli conda project run print-vars --version
+0.0.1
+```
+
```

Comment:
Can `conda project run` execute a named command in a chosen named environment selected on the command line? 

Just curious, not sure it is a good idea necessarily but logically it is something you might want to control from outside the YAML file...

Reply:
Yes run accepts `--environment` for both ad-hoc and defined commands and it will work as a way to switch the env for a defined command when executed.

Reply:
Was mentioning this to @philippjfr and he brought up an interesting point - can you also pick from the globally available conda environment names or just the ones specified in the yaml?

I think that would offer some interesting possibilities...

Reply:
Note that I'm not deeply attached to the notion that this should be possible. I've just been thinking about this in the context of an unnamed project, where, due to storage limitations, we will likely just use the global environment and re-create the exact environment embedded in the project.

Reply:
Do you mean like the anaconda-project `ANACONDA_PROJECT_ENVS_PATH` feature that can utilize global envs, but requires a matching spec in the project? Or something more like "I want to write a spec for conda-run to define commands, but I don't want the env management and locking features, I will just use the global conda envs to execute this"

https://anaconda-project.readthedocs.io/en/latest/config.html

Reply:
Personally, I was thinking of the latter case - using a project defined command with an arbitrary environment that is not managed by the project at all. It is then up to the user to pick an environment that has all the dependencies installed to support that command.

I see this possibility as an occasionally useful escape hatch and not as something I would recommend as regular good practice.

Reply:
https://github.com/conda-incubator/conda-project/issues/79

Reply:
I guess you might not expect to have to set an environment variable tbh: you would expect `conda project` to see whatever is available under `conda env list` (and not just what is in the yaml). However, I do *not* want this to be available by default, or at least behind a different flag (e.g  available when using`--global-environment` or similar).

Reply:
Thanks, let's continue on the issue